carcara/src/checker/rules/strings.rs

```diff
@@ -15,91 +15,162 @@ fn flatten(term: Rc<Term>, pool: &mut dyn TermPool) -> Vec<Rc<Term>> {
+            pool.add(Term::Let(binding.clone(), new_inner))
+        }
+        Term::Binder(q, bindings, inner) => {
+            let new_inner = expand_string_constants(pool, inner);
+            pool.add(Term::Binder(*q, bindings.clone(), new_inner))
+        }
+        Term::ParamOp { op, op_args, args } => {
+            let new_op_args = op_args
+                .iter()
+                .map(|term| expand_string_constants(pool, term))
+                .collect();
+            let new_args = args
+                .iter()
+                .map(|term| expand_string_constants(pool, term))
+                .collect();
+            pool.add(Term::ParamOp {
+                op: *op,
+                op_args: new_op_args,
+                args: new_args,
+            })
+        }
+        Term::Var(..) | Term::Const(_) | Term::Sort(_) => term.clone(),
+    }
+}
+
+fn concat(pool: &mut dyn TermPool, terms: Vec<Rc<Term>>) -> Rc<Term> {
     match terms.len() {
-        0 => pool.add(Term::Const(Constant::String("".to_owned()))),
+        0 => pool.add(Term::new_string("")),
         1 => terms[0].clone(),
         _ => pool.add(Term::Op(Operator::StrConcat, terms)),
     }
 }
 
-fn build_skolem_prefix(u: Rc<Term>, pool: &mut dyn TermPool) -> Rc<Term> {
-    build_term!(pool, (strsubstr {u.clone()} 0 (- (strlen {u.clone()}) 1)))
-}
-
-fn build_skolem_suffix(u: Rc<Term>, n: usize, pool: &mut dyn TermPool) -> Rc<Term> {
-    let n = pool.add(Term::new_int(n));
-    build_term!(pool, (strsubstr {u.clone()} {n.clone()} (- (strlen {u.clone()}) {n})))
-}
-
-fn is_prefix(
+fn is_prefix_or_suffix(
+    pool: &mut dyn TermPool,
     term: Rc<Term>,
     pref: Rc<Term>,
-    pool: &mut dyn TermPool,
     rev: bool,
     polyeq_time: &mut Duration,
 ) -> Result<Vec<Rc<Term>>, CheckerError> {
-    let mut t_flat = flatten(term.clone(), pool);
-    let mut p_flat = flatten(pref.clone(), pool);
-
+    let mut t_flat = flatten(pool, term.clone());
+    let mut p_flat = flatten(pool, pref.clone());
     if rev {
         t_flat.reverse();
         p_flat.reverse();
     }
-
     if p_flat.len() > t_flat.len() {
         return Err(CheckerError::ExpectedToBePrefix(pref, term));
     }
-
     for (i, el) in p_flat.iter().enumerate() {
         assert_polyeq_expected(el, t_flat[i].clone(), polyeq_time)?;
     }
-
+    if rev {
+        p_flat.reverse();
+    }
     Ok(p_flat)
 }
 
+fn is_prefix(
+    pool: &mut dyn TermPool,
+    term: Rc<Term>,
+    pref: Rc<Term>,
+    polyeq_time: &mut Duration,
+) -> Result<Vec<Rc<Term>>, CheckerError> {
+    is_prefix_or_suffix(pool, term, pref, false, polyeq_time)
+}
+
+fn is_suffix(
+    pool: &mut dyn TermPool,
+    term: Rc<Term>,
+    pref: Rc<Term>,
+    polyeq_time: &mut Duration,
+) -> Result<Vec<Rc<Term>>, CheckerError> {
+    is_prefix_or_suffix(pool, term, pref, true, polyeq_time)
+}
+
 type Suffixes = (Vec<Rc<Term>>, Vec<Rc<Term>>);
 
-fn strip_prefix(
+fn strip_prefix_or_suffix(
```

Comment:
It would be nice to have a comment documenting what this functions does. This is especially true for these helper functions that are used in multiple rules (e.g. `strip_prefix_or_suffix`, `flatten`, `expand_string_constants`, etc.).